src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs

```diff
@@ -215,6 +229,63 @@ private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, Co
             return false;
         }
 
+        // We should remove Razor design time helpers from C#'s completion list. If the current identifier being targeted does not start with a double
+        // underscore, we trim out all items starting with "__" from the completion list. If the current identifier does start with a double underscore
+        // (e.g. "__ab[||]"), we only trim out common design time helpers from the completion list.
+        private SumType<CompletionItem[], CompletionList>? RemoveDesignTimeItems(
+            SumType<CompletionItem[], CompletionList> completionResult,
+            LSPDocumentSnapshot documentSnapshot,
+            TextExtent? wordExtent,
+            LanguageServerKind serverKind)
+        {
+            if (serverKind != LanguageServerKind.CSharp)
+            {
+                return completionResult;
+            }
+
+            var result = completionResult.Match<SumType<CompletionItem[], CompletionList>?>(
+                items => RemoveDesignTimeItems(items, documentSnapshot, wordExtent),
+                list => RemoveDesignTimeItems(list.Items, documentSnapshot, wordExtent));
+
+            return result;
+
+            static CompletionItem[] RemoveDesignTimeItems(CompletionItem[] items, LSPDocumentSnapshot documentSnapshot, TextExtent? wordExtent)
+            {
+                var designTimeHelpersCompletionItems = GenerateCompletionItems(DesignTimeHelpers);
+                if (wordExtent.HasValue)
+                {
+                    var wordCharacterStartIndex = wordExtent.Value.Span.Start.Position;
+                    if (wordCharacterStartIndex + 1 < documentSnapshot.Snapshot.Length)
```

Comment:
Hmm, feels like an odd way to check if the word is at least 2 characters. Probably makes more sense to do
```C#
var wordSpan = wordExtent.Value.Span;
if (wordSpan.Length >= 2)
{
...
}
```